src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java

```diff
@@ -675,58 +621,30 @@ void blockingReceive(DatagramPacket p, long nanos) throws IOException {
             }
 
-            long startNanos = System.nanoTime();
-            long remainingNanos = nanos;
-            SocketAddress sender = null;
+            boolean completed = false;
             try {
                 SocketAddress remote = beginRead(true, false);
                 boolean connected = (remote != null);
-                do {
-                    ByteBuffer dst = tryBlockingReceive(connected, bufLength, remainingNanos);
-
+                ByteBuffer dst = tryBlockingReceive(connected, bufLength, nanos);
+                if (dst != null) {
                     // if datagram received then get sender and copy to DatagramPacket
-                    if (dst != null) {
-                        try {
-                            // sender address is in socket address buffer
-                            sender = sourceSocketAddress();
-
-                            // check sender when security manager set and not connected
-                            @SuppressWarnings("removal")
-                            SecurityManager sm = System.getSecurityManager();
-                            if (sm != null && !connected) {
-                                InetSocketAddress isa = (InetSocketAddress) sender;
-                                try {
-                                    sm.checkAccept(isa.getAddress().getHostAddress(), isa.getPort());
-                                } catch (SecurityException e) {
-                                    sender = null;
-                                }
-                            }
-
-                            if (sender != null) {
-                                // copy bytes to the DatagramPacket, and set length and sender
-                                synchronized (p) {
-                                    // re-read p.bufLength in case DatagramPacket changed
-                                    int len = Math.min(dst.limit(), DatagramPackets.getBufLength(p));
-                                    dst.get(p.getData(), p.getOffset(), len);
-                                    DatagramPackets.setLength(p, len);
-                                    p.setSocketAddress(sender);
-                                }
-                            } else {
-                                // need to retry, adjusting timeout if needed
-                                if (nanos > 0) {
-                                    remainingNanos = nanos - (System.nanoTime() - startNanos);
-                                    if (remainingNanos <= 0) {
-                                        throw new SocketTimeoutException("Receive timed out");
-                                    }
-                                }
-                            }
-                        } finally {
-                            Util.offerFirstTemporaryDirectBuffer(dst);
+                    try {
+                        SocketAddress sender = sourceSocketAddress();
+                        synchronized (p) {
```

Comment:
Not this PR but there is a follow-up coming that will drop the dependency on DatagramPacket from DatagramChannelImpl so that it does back to the adaptor.

Reply:
OK - if you have follow-ups in the pipe to fix this then good to go!

Reply:
Yes, there will be a follow-up from JEP 491 so that we can go back to using synchronization in the adaptor send/receive methods. I initially had everything in one patch but it's large so decided to split it and address the SM removal first.